File: packages/dart_node_react_native/test/ui.test.js
/**
 * UI Interaction Tests for Dart React Native Components
 * Uses React Native Testing Library for real component testing
 */

const { render, fireEvent, waitFor } = require('@testing-library/react-native');

// Make testing library available globally for Dart interop
// dart2js uses init.G which resolves to self || globalThis
global.rnTestingLibrary = { render, fireEvent, waitFor };
globalThis.rnTestingLibrary = global.rnTestingLibrary;

// Load compiled Dart code
require('./dist/ui_test.js');

// Get exported Dart components
const {
  counterComponent,
  formComponent,
  toggleComponent,
  listComponent,
  switchComponent,
  loadingComponent,
} = global.dartReactNativeTests;

// =============================================================================
// Counter Component Tests
// =============================================================================

describe('Counter Component', () => {
  test('renders with initial count of 0', () => {
    const { getByTestId } = render(counterComponent());

    const display = getByTestId('count-display');
    expect(display).toHaveTextContent('Count: 0');
  });

  test('increments count when increment button pressed', () => {
    const { getByTestId } = render(counterComponent());

    const incrementBtn = getByTestId('increment-btn');
    const display = getByTestId('count-display');

    fireEvent.press(incrementBtn);
    expect(display).toHaveTextContent('Count: 1');

    fireEvent.press(incrementBtn);
    expect(display).toHaveTextContent('Count: 2');

    fireEvent.press(incrementBtn);
    expect(display).toHaveTextContent('Count: 3');
  });

  test('decrements count when decrement button pressed', () => {
    const { getByTestId } = render(counterComponent());

    const decrementBtn = getByTestId('decrement-btn');
    const display = getByTestId('count-display');

    fireEvent.press(decrementBtn);
    expect(display).toHaveTextContent('Count: -1');

    fireEvent.press(decrementBtn);
    expect(display).toHaveTextContent('Count: -2');
  });

  test('resets count to 0 when reset button pressed', () => {
    const { getByTestId } = render(counterComponent());

    const incrementBtn = getByTestId('increment-btn');
    const resetBtn = getByTestId('reset-btn');
    const display = getByTestId('count-display');

    // Increment a few times
    fireEvent.press(incrementBtn);
    fireEvent.press(incrementBtn);
    fireEvent.press(incrementBtn);
    expect(display).toHaveTextContent('Count: 3');

    // Reset
    fireEvent.press(resetBtn);
    expect(display).toHaveTextContent('Count: 0');
  });

  test('handles rapid presses correctly', () => {
    const { getByTestId } = render(counterComponent());

    const incrementBtn = getByTestId('increment-btn');
    const display = getByTestId('count-display');

    // Rapid presses
    for (let i = 0; i < 10; i++) {
      fireEvent.press(incrementBtn);
    }

    expect(display).toHaveTextContent('Count: 10');
  });
});

// =============================================================================
// Form Component Tests
// =============================================================================

describe('Form Component', () => {
  test('renders empty form inputs', () => {
    const { getByTestId } = render(formComponent(null));

    const emailInput = getByTestId('email-input');
    const passwordInput = getByTestId('password-input');

    expect(emailInput.props.value).toBe('');
    expect(passwordInput.props.value).toBe('');
  });

  test('updates email input when typing', () => {
    const { getByTestId } = render(formComponent(null));

    const emailInput = getByTestId('email-input');

    fireEvent.changeText(emailInput, 'test@example.com');
    expect(emailInput.props.value).toBe('test@example.com');
  });

  test('updates password input when typing', () => {
    const { getByTestId } = render(formComponent(null));

    const passwordInput = getByTestId('password-input');

    fireEvent.changeText(passwordInput, 'secretpass123');
    expect(passwordInput.props.value).toBe('secretpass123');
  });

  test('shows error when submitting empty form', () => {
    const { getByTestId } = render(formComponent(null));

    const submitBtn = getByTestId('submit-btn');

    fireEvent.press(submitBtn);

    const errorMessage = getByTestId('error-message');
    expect(errorMessage).toHaveTextContent('All fields required');
  });

  test('shows error when only email is filled', () => {
    const { getByTestId } = render(formComponent(null));

    const emailInput = getByTestId('email-input');
    const submitBtn = getByTestId('submit-btn');

    fireEvent.changeText(emailInput, 'test@example.com');
    fireEvent.press(submitBtn);

    const errorMessage = getByTestId('error-message');
    expect(errorMessage).toHaveTextContent('All fields required');
  });

  test('shows error when only password is filled', () => {
    const { getByTestId } = render(formComponent(null));

    const passwordInput = getByTestId('password-input');
    const submitBtn = getByTestId('submit-btn');

    fireEvent.changeText(passwordInput, 'secret123');
    fireEvent.press(submitBtn);

    const errorMessage = getByTestId('error-message');
    expect(errorMessage).toHaveTextContent('All fields required');
  });

  test('shows success message when form is valid', () => {
    const { getByTestId, queryByTestId } = render(formComponent(null));

    const emailInput = getByTestId('email-input');
    const passwordInput = getByTestId('password-input');
    const submitBtn = getByTestId('submit-btn');

    fireEvent.changeText(emailInput, 'test@example.com');
    fireEvent.changeText(passwordInput, 'secret123');
    fireEvent.press(submitBtn);

    const successMessage = getByTestId('success-message');
    expect(successMessage).toHaveTextContent('Form submitted!');

    // Error should not be visible
    expect(queryByTestId('error-message')).toBeNull();
  });

  test('calls onSubmit callback with form values', () => {
    const onSubmit = jest.fn();

    const { getByTestId } = render(formComponent(onSubmit));

    const emailInput = getByTestId('email-input');
    const passwordInput = getByTestId('password-input');
    const submitBtn = getByTestId('submit-btn');

    fireEvent.changeText(emailInput, 'user@test.com');
    fireEvent.changeText(passwordInput, 'mypassword');
    fireEvent.press(submitBtn);

    expect(onSubmit).toHaveBeenCalledTimes(1);
    expect(onSubmit).toHaveBeenCalledWith('user@test.com', 'mypassword');
  });

  test('clears error when valid form is submitted', () => {
    const { getByTestId, queryByTestId } = render(formComponent(null));

    const emailInput = getByTestId('email-input');
    const passwordInput = getByTestId('password-input');
    const submitBtn = getByTestId('submit-btn');

    // First submit empty form to show error
    fireEvent.press(submitBtn);
    expect(getByTestId('error-message')).toBeTruthy();

    // Now fill and submit
    fireEvent.changeText(emailInput, 'test@example.com');
    fireEvent.changeText(passwordInput, 'password123');
    fireEvent.press(submitBtn);

    // Error should be gone, success should show
    expect(queryByTestId('error-message')).toBeNull();
    expect(getByTestId('success-message')).toBeTruthy();
  });
});

// =============================================================================
// Toggle Component Tests
// =============================================================================

describe('Toggle Component', () => {
  test('renders with OFF state initially', () => {
    const { getByTestId } = render(toggleComponent());

    const status = getByTestId('toggle-status');
    expect(status).toHaveTextContent('OFF');
  });

  test('toggles to ON when button pressed', () => {
    const { getByTestId } = render(toggleComponent());

    const toggleBtn = getByTestId('toggle-btn');
    const status = getByTestId('toggle-status');

    fireEvent.press(toggleBtn);
    expect(status).toHaveTextContent('ON');
  });

  test('toggles back to OFF when pressed again', () => {
    const { getByTestId } = render(toggleComponent());

    const toggleBtn = getByTestId('toggle-btn');
    const status = getByTestId('toggle-status');

    fireEvent.press(toggleBtn);
    expect(status).toHaveTextContent('ON');

    fireEvent.press(toggleBtn);
    expect(status).toHaveTextContent('OFF');
  });

  test('handles multiple toggles correctly', () => {
    const { getByTestId } = render(toggleComponent());

    const toggleBtn = getByTestId('toggle-btn');
    const status = getByTestId('toggle-status');

    // Toggle 5 times
    fireEvent.press(toggleBtn); // ON
    expect(status).toHaveTextContent('ON');

    fireEvent.press(toggleBtn); // OFF
    expect(status).toHaveTextContent('OFF');

    fireEvent.press(toggleBtn); // ON
    expect(status).toHaveTextContent('ON');

    fireEvent.press(toggleBtn); // OFF
    expect(status).toHaveTextContent('OFF');

    fireEvent.press(toggleBtn); // ON
    expect(status).toHaveTextContent('ON');
  });
});

// =============================================================================
// List Component Tests
// =============================================================================

describe('List Component', () => {
  test('renders empty list with message', () => {
    const { getByTestId } = render(listComponent());

    const emptyMessage = getByTestId('empty-message');
    expect(emptyMessage).toHaveTextContent('No items');

    const count = getByTestId('items-count');
    expect(count).toHaveTextContent('Total: 0');
  });

  test('adds item to list when add button pressed', () => {
    const { getByTestId, queryByTestId } = render(listComponent());

    const input = getByTestId('add-item-input');
    const addBtn = getByTestId('add-item-btn');

    fireEvent.changeText(input, 'First item');
    fireEvent.press(addBtn);

    const item = getByTestId('list-item-0');
    expect(item).toHaveTextContent('First item');

    const count = getByTestId('items-count');
    expect(count).toHaveTextContent('Total: 1');

    // Empty message should be gone
    expect(queryByTestId('empty-message')).toBeNull();
  });

  test('clears input after adding item', () => {
    const { getByTestId } = render(listComponent());

    const input = getByTestId('add-item-input');
    const addBtn = getByTestId('add-item-btn');

    fireEvent.changeText(input, 'Test item');
    fireEvent.press(addBtn);

    expect(input.props.value).toBe('');
  });

  test('does not add empty items', () => {
    const { getByTestId } = render(listComponent());

    const addBtn = getByTestId('add-item-btn');

    fireEvent.press(addBtn);

    // Should still show empty message
    expect(getByTestId('empty-message')).toBeTruthy();

    const count = getByTestId('items-count');
    expect(count).toHaveTextContent('Total: 0');
  });

  test('does not add whitespace-only items', () => {
    const { getByTestId } = render(listComponent());

    const input = getByTestId('add-item-input');
    const addBtn = getByTestId('add-item-btn');

    fireEvent.changeText(input, '   ');
    fireEvent.press(addBtn);

    expect(getByTestId('empty-message')).toBeTruthy();
  });

  test('adds multiple items correctly', () => {
    const { getByTestId } = render(listComponent());

    const input = getByTestId('add-item-input');
    const addBtn = getByTestId('add-item-btn');

    fireEvent.changeText(input, 'Item 1');
    fireEvent.press(addBtn);

    fireEvent.changeText(input, 'Item 2');
    fireEvent.press(addBtn);

    fireEvent.changeText(input, 'Item 3');
    fireEvent.press(addBtn);

    expect(getByTestId('list-item-0')).toHaveTextContent('Item 1');
    expect(getByTestId('list-item-1')).toHaveTextContent('Item 2');
    expect(getByTestId('list-item-2')).toHaveTextContent('Item 3');

    const count = getByTestId('items-count');
    expect(count).toHaveTextContent('Total: 3');
  });

  test('removes item when remove button pressed', () => {
    const { getByTestId, queryByTestId } = render(listComponent());

    const input = getByTestId('add-item-input');
    const addBtn = getByTestId('add-item-btn');

    fireEvent.changeText(input, 'Item to remove');
    fireEvent.press(addBtn);

    expect(getByTestId('list-item-0')).toBeTruthy();

    const removeBtn = getByTestId('remove-btn-0');
    fireEvent.press(removeBtn);

    expect(queryByTestId('list-item-0')).toBeNull();
    expect(getByTestId('empty-message')).toBeTruthy();
  });

  test('removes correct item from middle of list', () => {
    const { getByTestId, queryByTestId } = render(listComponent());

    const input = getByTestId('add-item-input');
    const addBtn = getByTestId('add-item-btn');

    fireEvent.changeText(input, 'First');
    fireEvent.press(addBtn);

    fireEvent.changeText(input, 'Second');
    fireEvent.press(addBtn);

    fireEvent.changeText(input, 'Third');
    fireEvent.press(addBtn);

    // Remove middle item
    const removeBtn = getByTestId('remove-btn-1');
    fireEvent.press(removeBtn);

    // Check remaining items
    expect(getByTestId('list-item-0')).toHaveTextContent('First');
    expect(getByTestId('list-item-1')).toHaveTextContent('Third');
    expect(queryByTestId('list-item-2')).toBeNull();

    const count = getByTestId('items-count');
    expect(count).toHaveTextContent('Total: 2');
  });

  test('handles add and remove operations in sequence', () => {
    const { getByTestId } = render(listComponent());

    const input = getByTestId('add-item-input');
    const addBtn = getByTestId('add-item-btn');

    // Add 3 items
    fireEvent.changeText(input, 'A');
    fireEvent.press(addBtn);
    fireEvent.changeText(input, 'B');
    fireEvent.press(addBtn);
    fireEvent.changeText(input, 'C');
    fireEvent.press(addBtn);

    expect(getByTestId('items-count')).toHaveTextContent('Total: 3');

    // Remove first
    fireEvent.press(getByTestId('remove-btn-0'));
    expect(getByTestId('items-count')).toHaveTextContent('Total: 2');

    // Add new item
    fireEvent.changeText(input, 'D');
    fireEvent.press(addBtn);
    expect(getByTestId('items-count')).toHaveTextContent('Total: 3');

    // Remove all
    fireEvent.press(getByTestId('remove-btn-0'));
    fireEvent.press(getByTestId('remove-btn-0'));
    fireEvent.press(getByTestId('remove-btn-0'));

    expect(getByTestId('empty-message')).toBeTruthy();
    expect(getByTestId('items-count')).toHaveTextContent('Total: 0');
  });
});

// =============================================================================
// Switch Component Tests
// =============================================================================

describe('Switch Component', () => {
  test('renders with Disabled state initially', () => {
    const { getByTestId } = render(switchComponent());

    const status = getByTestId('switch-status');
    expect(status).toHaveTextContent('Disabled');
  });

  test('switch toggles state when value changes', () => {
    const { getByTestId } = render(switchComponent());

    const switchControl = getByTestId('switch-control');
    const status = getByTestId('switch-status');

    // Simulate switch toggle
    fireEvent(switchControl, 'valueChange', true);
    expect(status).toHaveTextContent('Enabled');

    fireEvent(switchControl, 'valueChange', false);
    expect(status).toHaveTextContent('Disabled');
  });
});

// =============================================================================
// Loading Component Tests
// =============================================================================

describe('Loading Component', () => {
  test('renders loading indicator initially', () => {
    const { getByTestId, queryByTestId } = render(loadingComponent());

    const indicator = getByTestId('loading-indicator');
    expect(indicator).toBeTruthy();

    expect(queryByTestId('loaded-content')).toBeNull();
  });

  test('shows content when loading is finished', () => {
    const { getByTestId, queryByTestId } = render(loadingComponent());

    const toggleBtn = getByTestId('toggle-loading-btn');

    fireEvent.press(toggleBtn);

    expect(queryByTestId('loading-indicator')).toBeNull();
    expect(getByTestId('loaded-content')).toHaveTextContent('Content loaded');
  });

  test('toggles back to loading state', () => {
    const { getByTestId, queryByTestId } = render(loadingComponent());

    const toggleBtn = getByTestId('toggle-loading-btn');

    // Finish loading
    fireEvent.press(toggleBtn);
    expect(getByTestId('loaded-content')).toBeTruthy();

    // Start loading again
    fireEvent.press(toggleBtn);
    expect(getByTestId('loading-indicator')).toBeTruthy();
    expect(queryByTestId('loaded-content')).toBeNull();
  });
});

File: packages/dart_node_react_native/test/ui_test.dart
/// UI interaction tests for React Native components
/// Compiled to JS and run with Jest + React Native Testing Library
///
/// Build: dart compile js test/ui_test.dart -o test/dist/ui_test.js
/// Run: cd test && npm test
library;

import 'dart:js_interop';
import 'dart:js_interop_unsafe';

import 'package:dart_node_react_native/dart_node_react_native.dart';

// =============================================================================
// Testing Library Bindings
// =============================================================================

@JS('rnTestingLibrary.render')
external JSObject _render(JSObject element);

@JS('rnTestingLibrary.screen')
external JSObject get _screen;

@JS('rnTestingLibrary.fireEvent')
external JSObject get _fireEvent;

@JS('rnTestingLibrary.waitFor')
external JSPromise<JSAny?> _waitFor(JSFunction callback);

// =============================================================================
// Test Utilities
// =============================================================================

/// Render a React Native element for testing
JSObject render(JSObject element) => _render(element);

/// Screen queries
JSObject get screen => _screen;

/// Fire events on elements
JSObject get fireEvent => _fireEvent;

/// Wait for async operations - takes a JSFunction callback
/// Returns a JSPromise that can be awaited in JS
JSPromise<JSAny?> waitFor(JSFunction callback) => _waitFor(callback);

/// Query by test ID
JSObject? queryByTestId(String testId) =>
    screen.callMethod('queryByTestId'.toJS, testId.toJS) as JSObject?;

/// Get by test ID (throws if not found)
JSObject getByTestId(String testId) =>
    switch (screen.callMethod('getByTestId'.toJS, testId.toJS)) {
      final JSObject obj => obj,
      _ => throw StateError('Element with testId "$testId" not found'),
    };

/// Query by text
JSObject? queryByText(String text) =>
    screen.callMethod('queryByText'.toJS, text.toJS) as JSObject?;

/// Get by text (throws if not found)
JSObject getByText(String text) =>
    switch (screen.callMethod('getByText'.toJS, text.toJS)) {
      final JSObject obj => obj,
      _ => throw StateError('Element with text "$text" not found'),
    };

/// Get all by text
JSArray getAllByText(String text) =>
    switch (screen.callMethod('getAllByText'.toJS, text.toJS)) {
      final JSArray arr => arr,
      _ => throw StateError('No elements found with text "$text"'),
    };

/// Query by placeholder
JSObject? queryByPlaceholder(String text) =>
    screen.callMethod('queryByPlaceholderText'.toJS, text.toJS) as JSObject?;

/// Get by placeholder (throws if not found)
JSObject getByPlaceholder(String text) =>
    switch (screen.callMethod('getByPlaceholderText'.toJS, text.toJS)) {
      final JSObject obj => obj,
      _ => throw StateError('Element with placeholder "$text" not found'),
    };

/// Press on a touchable element
void press(JSObject element) {
  fireEvent.callMethod('press'.toJS, element);
}

/// Change text in a TextInput
void changeText(JSObject element, String text) {
  fireEvent.callMethod('changeText'.toJS, element, text.toJS);
}

// =============================================================================
// Test Components
// =============================================================================

/// Simple counter component for testing state updates
JSObject counterComponent() => functionalComponent('Counter', (props) {
      final (countState, setCount) = useState(0.toJS);
      final count = (countState as JSNumber?)?.toDartInt ?? 0;

      return view(
        props: {'testID': 'counter-container'},
        children: [
          text('Count: $count', props: {'testID': 'count-display'}),
          touchableOpacity(
            onPress: () => setCount.callAsFunction(null, (count + 1).toJS),
            props: {'testID': 'increment-btn'},
            child: text('Increment'),
          ),
          touchableOpacity(
            onPress: () => setCount.callAsFunction(null, (count - 1).toJS),
            props: {'testID': 'decrement-btn'},
            child: text('Decrement'),
          ),
          touchableOpacity(
            onPress: () => setCount.callAsFunction(null, 0.toJS),
            props: {'testID': 'reset-btn'},
            child: text('Reset'),
          ),
        ],
      );
    });

/// Form component for testing input interactions
JSObject formComponent(
        {void Function(String email, String password)? onSubmit}) =>
    functionalComponent('Form', (props) {
      final (emailState, setEmail) = useState(''.toJS);
      final (passState, setPass) = useState(''.toJS);
      final (errorState, setError) = useState(null);
      final (submittedState, setSubmitted) = useState(false.toJS);

      final email = (emailState as JSString?)?.toDart ?? '';
      final password = (passState as JSString?)?.toDart ?? '';
      final error = (errorState as JSString?)?.toDart;
      final submitted = (submittedState as JSBoolean?)?.toDart ?? false;

      void handleSubmit() {
        (email.isEmpty || password.isEmpty)
            ? setError.callAsFunction(null, 'All fields required'.toJS)
            : () {
                setError.callAsFunction();
                setSubmitted.callAsFunction(null, true.toJS);
                onSubmit?.call(email, password);
              }();
      }

      return view(
        props: {'testID': 'form-container'},
        children: [
          if (error != null)
            view(
              props: {'testID': 'error-message'},
              child: text(error),
            )
          else
            view(),
          if (submitted)
            view(
              props: {'testID': 'success-message'},
              child: text('Form submitted!'),
            )
          else
            view(),
          textInput(
            placeholder: 'Enter email',
            value: email,
            onChangeText: (val) => setEmail.callAsFunction(null, val.toJS),
            props: {'testID': 'email-input'},
          ),
          textInput(
            placeholder: 'Enter password',
            value: password,
            secureTextEntry: true,
            onChangeText: (val) => setPass.callAsFunction(null, val.toJS),
            props: {'testID': 'password-input'},
          ),
          touchableOpacity(
            onPress: handleSubmit,
            props: {'testID': 'submit-btn'},
            child: text('Submit'),
          ),
        ],
      );
    });

/// Toggle component for testing boolean state
JSObject toggleComponent() => functionalComponent('Toggle', (props) {
      final (onState, setOn) = useState(false.toJS);
      final isOn = (onState as JSBoolean?)?.toDart ?? false;

      return view(
        props: {'testID': 'toggle-container'},
        children: [
          text(
            isOn ? 'ON' : 'OFF',
            props: {'testID': 'toggle-status'},
          ),
          touchableOpacity(
            onPress: () => setOn.callAsFunction(null, (!isOn).toJS),
            props: {'testID': 'toggle-btn'},
            child: text('Toggle'),
          ),
        ],
      );
    });

/// List component for testing dynamic rendering
JSObject listComponent() => functionalComponent('List', (props) {
      final (itemsState, setItems) = useState(<JSAny>[].toJS);
      final (inputState, setInput) = useState(''.toJS);

      final items = (itemsState as JSArray?)?.toDart ?? [];
      final inputValue = (inputState as JSString?)?.toDart ?? '';

      void addItem() => switch (inputValue.trim().isNotEmpty) {
            true => () {
                final newItems = [...items, inputValue.toJS];
                setItems.callAsFunction(null, newItems.toJS);
                setInput.callAsFunction(null, ''.toJS);
              }(),
            false => null,
          };

      void removeItem(int index) {
        final newItems = [...items]..removeAt(index);
        setItems.callAsFunction(null, newItems.toJS);
      }

      return view(
        props: {'testID': 'list-container'},
        children: [
          view(children: [
            textInput(
              placeholder: 'Add item',
              value: inputValue,
              onChangeText: (val) => setInput.callAsFunction(null, val.toJS),
              props: {'testID': 'add-item-input'},
            ),
            touchableOpacity(
              onPress: addItem,
              props: {'testID': 'add-item-btn'},
              child: text('Add'),
            ),
          ]),
          view(
            props: {'testID': 'items-list'},
            children: items.isEmpty
                ? [
                    text('No items', props: {'testID': 'empty-message'})
                  ]
                : items
                    .asMap()
                    .entries
                    .map(
                      (entry) => view(
                        props: {
                          'key': entry.key,
                          'testID': 'list-item-${entry.key}'
                        },
                        children: [
                          text(
                            switch (entry.value) {
                              final JSString s => s.toDart,
                              _ => '',
                            },
                          ),
                          touchableOpacity(
                            onPress: () => removeItem(entry.key),
                            props: {'testID': 'remove-btn-${entry.key}'},
                            child: text('Remove'),
                          ),
                        ],
                      ),
                    )
                    .toList(),
          ),
          text(
            'Total: ${items.length}',
            props: {'testID': 'items-count'},
          ),
        ],
      );
    });

/// Switch component for testing toggle with RN Switch
JSObject switchComponent() => functionalComponent('SwitchTest', (props) {
      final (onState, setOn) = useState(false.toJS);
      final isOn = (onState as JSBoolean?)?.toDart ?? false;

      return view(
        props: {'testID': 'switch-container'},
        children: [
          text(
            isOn ? 'Enabled' : 'Disabled',
            props: {'testID': 'switch-status'},
          ),
          rnSwitch(
            value: isOn,
            onValueChange: (val) => setOn.callAsFunction(null, val.toJS),
            props: {'testID': 'switch-control'},
          ),
        ],
      );
    });

/// Activity indicator component for testing loading states
JSObject loadingComponent() => functionalComponent('Loading', (props) {
      final (loadingState, setLoading) = useState(true.toJS);
      final isLoading = (loadingState as JSBoolean?)?.toDart ?? true;

      return view(
        props: {'testID': 'loading-container'},
        children: [
          if (isLoading)
            activityIndicator(
              size: 'large',
              color: '#6366f1',
              props: {'testID': 'loading-indicator'},
            )
          else
            text('Content loaded', props: {'testID': 'loaded-content'}),
          touchableOpacity(
            onPress: () => setLoading.callAsFunction(null, (!isLoading).toJS),
            props: {'testID': 'toggle-loading-btn'},
            child: text(isLoading ? 'Finish loading' : 'Start loading'),
          ),
        ],
      );
    });

// =============================================================================
// Export test components for Jest
// =============================================================================

@JS('dartReactNativeTests')
external set _dartReactNativeTests(JSObject value);

void main() {
  _dartReactNativeTests = JSObject()
    // Export components
    ..setProperty('counterComponent'.toJS, counterComponent.toJS)
    ..setProperty(
        'formComponent'.toJS,
        ((JSAny? onSubmit) => (onSubmit == null || onSubmit.isUndefinedOrNull)
            ? formComponent()
            : formComponent(
                onSubmit: (email, password) => (onSubmit as JSFunction)
                    .callAsFunction(null, email.toJS, password.toJS),
              )).toJS)
    ..setProperty('toggleComponent'.toJS, toggleComponent.toJS)
    ..setProperty('listComponent'.toJS, listComponent.toJS)
    ..setProperty('switchComponent'.toJS, switchComponent.toJS)
    ..setProperty('loadingComponent'.toJS, loadingComponent.toJS)
    // Export utilities
    ..setProperty('render'.toJS, render.toJS)
    ..setProperty('fireEvent'.toJS, fireEvent)
    ..setProperty('getByTestId'.toJS, getByTestId.toJS)
    ..setProperty('getByPlaceholder'.toJS, getByPlaceholder.toJS)
    ..setProperty('queryByText'.toJS, queryByText.toJS)
    ..setProperty('getByText'.toJS, getByText.toJS)
    ..setProperty('getAllByText'.toJS, getAllByText.toJS)
    ..setProperty('press'.toJS, press.toJS)
    ..setProperty('queryByPlaceholder'.toJS, queryByPlaceholder.toJS)
    ..setProperty('queryByTestId'.toJS, queryByTestId.toJS)
    ..setProperty('changeText'.toJS, changeText.toJS)
    ..setProperty('waitFor'.toJS, waitFor.toJS);
}
